tools/scylla-sstable: add `scylla sstable shard-of` command

tchaikov opened the pull request:
when migrating to the uuid-based identifiers, the mapping from the integer-based generation to the shard-id is preserved. we used to have "gen % smp_count" for calculating the shard which is responsible to host a given sstable. despite that this is not a documented behavior, this is handy when we try to correlate an sstable to a shard, typically when looking at a performance issue.

in this change, a new subcommand is added to expose the connection between the sstable and its "owner" shards.

Fixes #16343
Signed-off-by: Kefu Chai <kefu.chai@scylladb.com>

denesb commented on tools/scylla-sstable.cc:
```diff
@@ -2816,6 +2833,11 @@ for more information on this operation, including the API documentation.
+    {{"shard-of",
+            "Print out the shard which 'owns' the sstable",
+            "Print out the intersection(s) of the shard-ranges and the partition ranges"},
+            shard_of_operation},
```
I think we should have a `--text` (or --human-readeable or similar) option to output a human-friendly text output. The JSON output is hard to read without `jq` and the latter is not installed on every server.
I think we should go back and do the same with `validate` and `validate-checksums`, but that is for another PR.

denesb replied:
Maybe `--text|-t` is best, as it is short to type.

tchaikov replied:
> The JSON output is hard to read without `jq` and the latter is not installed on every server.

this applies to all subcommand which prints JSON. probably we should instead redistribute `jq` in the scylla image?

tchaikov replied:
i think this deserves a github issue.

avikivity replied:
maybe yaml output is more readable (all json can be printed as yaml)

tchaikov replied:
tracked at #16361

tchaikov replied:
i will add the `-t|--text` option in a follow-up change.

denesb commented on tools/scylla-sstable.cc:
```diff
@@ -2553,6 +2553,23 @@ void sstable_consumer_operation(schema_ptr schema, reader_permit permit, const s
+        writer.StartArray();
+        for (unsigned shard_id : sst->get_shards_for_this_sstable()) {
+            writer.Uint(shard_id);
+        }
```
Shouldn't this depend on the number of shards the source server has? I think this will work with sharder of the reconstructed schema which might be using smp=1.

tchaikov replied:
> Shouldn't this depend on the number of shards the source server has?

yes. but we don't encode the number of shards the source server has in the `Scylla` component, i think. if it's not there, could you please shed some lights on where can i find it?

> I think this will work with sharder of the reconstructed schema which might be using smp=1.

by default, the reconstructed schema uses the `smp::count` for creating its sharder, `--smp` is 1 if not specified. probably, i should add a note in the usage document to explain that user is supposed to specify `--smp` to get the correct output?

denesb replied:
We need a new `--shards` command-line parameter and make it required. Without this, the output of this command makes no sense.
I don't want to use `--smp`, in fact in the future I want to hide all but a few hand-selected seastar parameter when using the tools.

tchaikov replied:
ack. will add this option tomorrow.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5131/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6509/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5131/)
#### Build Details:
- Duration: 2 hr 40 min
- Builder: spider4.cloudius-systems.com

denesb commented:
Missing: tests.

tchaikov commented:
> Missing: tests.

would a dtest-based on suffice? if yes, i will add one .

denesb commented:
> > Missing: tests.
> 
> would a dtest-based on suffice? if yes, i will add one .

We have tests for scylla-sstable in tests/cql-pytest/test_tools.py. Please add a new test there. The scylla-instance started by `cql-pytest` runs with `--smp=2`, so you can create a table, write a partition belonging to 1 shard, flush, then repeat the same for the other shard, and then verify the sstables with `scylla sstable shardof`.
You can use `scylla types shardof` to check what shad a given partition-key belongs to. It is a little clunky to use, but you can look for examples in that file.

tchaikov commented:
thank you @denesb . in-tree tests are even simpler to hack / contribute.

tchaikov commented:
v2:

- added "--shards" option

i need to take more time to figure out how to "write a partition belonging to 1 shard".

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5203/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6595/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5203/)
#### Build Details:
- Duration: 2 hr 4 min
- Builder: spider6.cloudius-systems.com

avikivity commented:
> v2:
> 
> * added "--shards" option
> 
> i need to take more time to figure out how to "write a partition belonging to 1 shard".

Search for `local_shard_only`

tchaikov commented:
> > v2:
> > 
> > * added "--shards" option
> > 
> > i need to take more time to figure out how to "write a partition belonging to 1 shard".
> 
> Search for `local_shard_only`

thank you @avikivity , seems i need to create yet another command to expose this facility, so that we can use it in `tests/cql-pytest/test_tools.py` for preparing the test data set.

denesb commented:
@tchaikov it is also fine to write say 10 random partitions, flush, and check that the sstable belongs to the correct shard (you can check which shard the partition itself belongs to with `scylla types shardof`.

denesb commented:
> > v2:
> > 
> > * added "--shards" option
> > 
> > i need to take more time to figure out how to "write a partition belonging to 1 shard".
> 
> Search for `local_shard_only`

This is a C++ function and I asked @tchaikov to write a test in cql-pytest. This function simply keeps generating random keys until it finds one that belongs to the local shard. The same approach can be used in clq-pytest, by generating random keys and checking with `scylla types shardof`, but like I said above, this might not be necessary.

tchaikov commented:
> @tchaikov it is also fine to write say 10 random partitions, flush, and check that the sstable belongs to the correct shard (you can check which shard the partition itself belongs to with `scylla types shardof`.

@denesb thanks. but what is the source of truth regarding the "correct shard" in the Python land? unless i hardcode some known dataset, i am not sure how to do this in a randomized way in Python.

> The same approach can be used in clq-pytest, by generating random keys and checking with scylla types shardof, but like I said above, this might not be necessary.

without the function with tells if "(a,b,c)" goes to shard.x, it's impossible to use the same technique to create such a generator.

tchaikov commented:
v3:

- added a test for the newly introduced `shard-of` subcommand

@denesb hi Botond, i added a test to `cql-pytest/test_tools.py` as suggested. could you please take another look?

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5227/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6624/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5227/)
#### Failed Tests (12/23450):
- [test_scylla_sstable_shard_of](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5227/testReport/junit/%28root%29/test_tools/Tests___Unit_Tests___test_scylla_sstable_shard_of) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scylla_sstable_shard_of)
- [test_scylla_sstable_shard_of](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5227/testReport/junit/%28root%29/test_tools/Tests___Unit_Tests___test_scylla_sstable_shard_of_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scylla_sstable_shard_of)
- [test_scylla_sstable_shard_of](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5227/testReport/junit/%28root%29/test_tools/Tests___Unit_Tests___test_scylla_sstable_shard_of_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scylla_sstable_shard_of)
- [test_scylla_sstable_shard_of](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5227/testReport/junit/%28root%29/test_tools/Tests___Unit_Tests___test_scylla_sstable_shard_of_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scylla_sstable_shard_of)
- [test_scylla_sstable_shard_of](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5227/testReport/junit/%28root%29/test_tools/Tests___Unit_Tests___test_scylla_sstable_shard_of_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scylla_sstable_shard_of)
- [test_scylla_sstable_shard_of](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5227/testReport/junit/%28root%29/test_tools/Tests___Unit_Tests___test_scylla_sstable_shard_of_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scylla_sstable_shard_of)
- [cql-pytest.test_tools.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5227/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_pytest_test_tools_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.test_tools.dev.1)
- [cql-pytest.test_tools.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5227/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_pytest_test_tools_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.test_tools.dev.2)
- [cql-pytest.test_tools.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5227/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_pytest_test_tools_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.test_tools.dev.3)
- [cql-pytest.test_tools.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5227/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_pytest_test_tools_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.test_tools.release.1)
#### Build Details:
- Duration: 2 hr 59 min
- Builder: i-036fc42b48b85bbdb (m5d.12xlarge)

tchaikov commented:
```
[2023-12-13T17:22:26.197Z] .Segmentation fault.
[2023-12-13T17:22:26.197Z] Backtrace:
[2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3c594c
[2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3f1d96
[2023-12-13T17:22:26.197Z]   /lib64/libc.so.6+0x3dbaf
[2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x37132d
[2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x372983
[2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3c2fe4
[2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3dddca
[2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3f234c
[2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x38c4da
[2023-12-13T17:22:26.197Z]   /lib64/libc.so.6+0x8c946
[2023-12-13T17:22:26.197Z]   /lib64/libc.so.6+0x11285f
[2023-12-13T17:22:26.197Z] F
...
[2023-12-13T17:22:26.201Z] >               raise CalledProcessError(retcode, process.args,
[2023-12-13T17:22:26.201Z]                                          output=stdout, stderr=stderr)
[2023-12-13T17:22:26.201Z] E               subprocess.CalledProcessError: Command '['/jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/scylla', 'types', 'shardof', '--full-compound', '--shards', '2', '-t', 'Int32Type', '--', '000400000000']' died with <Signals.SIGSEGV: 11>.
[2023-12-13T17:22:26.201Z] 
[2023-12-13T17:22:26.201Z] /usr/lib64/python3.11/subprocess.py:571: CalledProcessError
```

only found in dev and release builds.

tchaikov commented:
> ```
> [2023-12-13T17:22:26.197Z] .Segmentation fault.
> [2023-12-13T17:22:26.197Z] Backtrace:
> [2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3c594c
> [2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3f1d96
> [2023-12-13T17:22:26.197Z]   /lib64/libc.so.6+0x3dbaf
> [2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x37132d
> [2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x372983
> [2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3c2fe4
> [2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3dddca
> [2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x3f234c
> [2023-12-13T17:22:26.197Z]   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/libseastar.so+0x38c4da
> [2023-12-13T17:22:26.197Z]   /lib64/libc.so.6+0x8c946
> [2023-12-13T17:22:26.197Z]   /lib64/libc.so.6+0x11285f
> [2023-12-13T17:22:26.197Z] F
> ...
> [2023-12-13T17:22:26.201Z] >               raise CalledProcessError(retcode, process.args,
> [2023-12-13T17:22:26.201Z]                                          output=stdout, stderr=stderr)
> [2023-12-13T17:22:26.201Z] E               subprocess.CalledProcessError: Command '['/jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/scylla', 'types', 'shardof', '--full-compound', '--shards', '2', '-t', 'Int32Type', '--', '000400000000']' died with <Signals.SIGSEGV: 11>.
> [2023-12-13T17:22:26.201Z] 
> [2023-12-13T17:22:26.201Z] /usr/lib64/python3.11/subprocess.py:571: CalledProcessError
> ```
> 
> only found in dev and release builds.

should be addressed by https://github.com/scylladb/seastar/pull/2000

denesb commented on tools/utils.cc:
```diff
@@ -111,6 +126,10 @@ int tool_app_template::run_async(int argc, char** argv, noncopyable_function<int
+    // i have to set the smp option before starting the reactor
+    if (unsigned shards = parse_shards_option(argc, argv); shards > 0) {
+        app_cfg.smp_opts.smp.set_value(shards);
+    }
```
This is too intrusive. We don't want to run the reactor with X shards just because we want to check an sstable with X shards. What if I run this on my laptop with 4 CPUs and I want to examine an sstable from a node with 64 CPUs?

There is a much simpler way: take the schema of the sstable and set a custom murmur sharder on it, configured with the desired shard count. Then set the schema of the sstable to this custom schema:
```c++
auto schema = sst->schema();
auto new_schema = schema_builder(schema).with_sharder(shard_count, sharding_ignore_msb_bits); // need to also expose an option for sharding_ignore_msb_bits, defaulted to 12 (which is the same default scylla config has)
```
Now we have the schema with the correct sharder, but the problem is that the sstable already loaded the sharding info with the default sharder (with `smp::count` shards).
There is no method to set a new schema. I think our best option is to reopen sstables with the new schema:
```c++
auto new_sst = sstable_manager.make_sstable(.../*derive these from existing sstable*/);
co_await sst->load(...);
// now print sharding info
```

denesb commented on test/cql-pytest/test_tools.py:
```diff
@@ -865,3 +868,83 @@ def test_scrub_validate_mode(scylla_path, scrub_workdir, scrub_schema_file, scru
+        cql.execute(f"INSERT INTO {keyspace}.{table} (pk, v) VALUES ({pk}, 0)")
+    nodetool.flush(cql, f"{keyspace}.{table}")
+
+    return table, schema
```
You need to clean up the table after the test, the scylla instance is shared by all tests.

You can use the pre-existing `new_test_table` context manager.

tchaikov replied:
i think `scylla_sstable()` does this for us already.

denesb replied:
Indeed.

denesb commented on test/cql-pytest/test_tools.py:
```diff
@@ -865,3 +868,83 @@ def test_scrub_validate_mode(scylla_path, scrub_workdir, scrub_schema_file, scru
+def _to_cql3_type(t: Type) -> str:
+    # map from Python type to Cassandra type, only a small subset is supported
+    py_to_cql3_type = {int: "Int32Type",
+                       str: "AsciiType",
```
UTF8Type (this is what the `text` cql type translates to).

tchaikov replied:
okay. changed as suggested.

i wanted to demonstrate an extensive way to build the command line. probably a better way is to use dedicate types as the python driver does to differentiate `UTF8Type` from `AsciiType`.  let's do this when necessary.

tchaikov commented:
v3 and v4:

- instead of setting `smp:count` for a shader with expected `shard_count`, reload the sstable instead
- s/AsciiType/UTF8Type/
- remove unused `typing.Any`

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5246/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6648/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5246/)
#### Build Details:
- Duration: 3 hr 13 min
- Builder: i-094ce1a6646595dea (m5d.12xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5251/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6655/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5251/)
#### Failed Tests (2/23451):
- [test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5251/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode)
- [topology_experimental_raft.test_tablets.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5251/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.debug.2)
#### Build Details:
- Duration: 2 hr 23 min
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5261/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6662/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5261/)
#### Build Details:
- Duration: 2 hr 4 min
- Builder: spider6.cloudius-systems.com

tchaikov commented:
> #### 🔴 CI State: ABORTED
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5246/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6648/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5246/)
> 
> #### Build Details:
> * Duration: 3 hr 13 min
> * Builder: i-094ce1a6646595dea (m5d.12xlarge)

the dtest aborted. the tests performed in this run passed.

tchaikov commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5251/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6655/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5251/)
> 
> #### Failed Tests (2/23451):
> * [test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5251/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode)
> * [topology_experimental_raft.test_tablets.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5251/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.debug.2)
> 

these are known issues. they are tracked by #14746

> #### Build Details:
> * Duration: 2 hr 23 min
> * Builder: spider4.cloudius-systems.com

tchaikov commented:
@scylladb/scylla-maint hello maintainers, could you help queue this change?

raphaelsc commented:
fast forward to tablets, we'll probably need to have a tablet count flag (to use tablet sharder instead), assuming we'll keep the power-of-two invariant (otherwise we can't make it offline AFAICT), so we can at least provide owner tablet id(s) from a given sstable's token range.

avikivity commented:
The tablet number doesn't say much since it says nothing about the shard.

raphaelsc commented:
> The tablet number doesn't say much since it says nothing about the shard.

AFAIU, we want this to help offline tools that group sstables by their owner. We don't necessarily have to reuse shard-of for that, and it actually makes sense to restrict it to vnodes (and deprecate it once we deprecate vnodes) and have a new sstable tablet-id-of for tablets.

raphaelsc commented:
But the world with tablets is too dynamic that it makes it much harder to build such offline tools. You run now with tablet count X, and coordinator decides to split / merge, and your output is outdated. Even though it's not perfect, it's better than nothing.

raphaelsc commented:
Migration doesn't matter, as we all we care is tablet id to perform the grouping. And at the end of the day, it's not like we'll be splitting or merging every minute. I think it might even make sense to have an http API to query tablet count for a tablet, without a CQL connection.

raphaelsc commented:
s/for a tablet/for a table/
